### PR TITLE
store, executor: support stale read for tikv RPCContext

### PR DESCRIPTION
**Pull request by Yisaer**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/21094

### What is changed and how it works?

Support stale read for tikv rpcContext

### Check List 

Tests 

- Unit test

### Release note 

* No release note

### CODE REVIEW

**Review by xhebox — COMMENTED** [review]
Though I think that you can avoid allocation for `[]StoreSelectorOption` by using an array or two branches, it is fine.

I will give a LGTM after passing all tests/upgrading protobuf stuff.

**Review by xhebox — COMMENTED** [review]
plz resolve the conflicts

File: store/tikv/tikvrpc/tikvrpc.go [hunk 1]
```diff
@@ -389,6 +389,13 @@ func (req *Request) TxnHeartBeat() *kvrpcpb.TxnHeartBeatRequest {
 	return req.Req.(*kvrpcpb.TxnHeartBeatRequest)
 }
 
+// EnableStaleRead enable the stale read
```
Comment (djshow832) [inline store/tikv/tikvrpc/tikvrpc.go:392]:
```suggestion
// EnableStaleRead enables stale read
```

File: store/tikv/tikvrpc/tikvrpc.go [hunk 2]
```diff
@@ -389,6 +389,13 @@ func (req *Request) TxnHeartBeat() *kvrpcpb.TxnHeartBeatRequest {
...
+func (req *Request) EnableStaleRead() {
+	req.StaleRead = true
+	req.ReplicaReadType = kv.ReplicaReadMixed
```
Comment (djshow832) [inline store/tikv/tikvrpc/tikvrpc.go:395]: Now that `req.ReplicaRead` is false, why set `req.ReplicaReadType`?

File: distsql/request_builder.go [hunk 3]
```diff
@@ -235,6 +236,15 @@ func (builder *RequestBuilder) SetFromSessionVars(sv *variable.SessionVars) *Req
 		builder.Request.SchemaVar = sv.TxnCtx.SchemaVersion
 	}
 	builder.txnScope = sv.TxnCtx.TxnScope
+	builder.IsStaleness = sv.TxnCtx.IsStaleness
+	if builder.IsStaleness && builder.txnScope != oracle.GlobalTxnScope {
```
Comment (djshow832) [inline distsql/request_builder.go:240]: Why doesn't global scope support stale read?

File: store/tikv/client_helper.go [hunk 4]
```diff
@@ -73,13 +74,22 @@ func (ch *ClientHelper) ResolveLocks(bo *Backoffer, callerStartTS uint64, locks
...
 	sender.Stats = ch.Stats
 	req.Context.ResolvedLocks = ch.resolvedLocks.GetAll()
-	resp, ctx, err := sender.SendReqCtx(bo, req, regionID, timeout, sType)
+	failpoint.Inject("assertStaleReadFlag", func(val failpoint.Value) {
+		if val.(bool) {
+			if len(opts) > 0 {
+				if !req.StaleRead {
+					panic("req.StaleRead should be true")
+				}
+			}
```
Comment (djshow832) [inline store/tikv/client_helper.go:90]: How about the situation that `opts` must be empty when `StaleRead` is false?

File: store/tikv/tikvrpc/tikvrpc.go (hunk 2 above)
Comment (Yisaer) [reply store/tikv/tikvrpc/tikvrpc.go:395]: `ReplicaReadType` is used for picking `store` in `GetTiKVRPCContext`.

File: distsql/request_builder.go (hunk 3 above)
Comment (Yisaer) [reply distsql/request_builder.go:240]: `global scope` also support stale read. If the txnScope is `local` here, we should pick `store` which match the label.

File: store/tikv/client_helper.go (hunk 4 above)
Comment (Yisaer) [reply store/tikv/client_helper.go]: I think this assertion is equal to your situation.

File: distsql/request_builder.go (hunk 3 above)
Comment (djshow832) [reply distsql/request_builder.go:240]: So is it possible to also replace `build.txnScope` with `builder.MatchStoreLabels`?

File: store/tikv/client_helper.go (hunk 4 above)
Comment (djshow832) [reply store/tikv/client_helper.go]: Your case is `len(opts) > 0 && !req.StaleRead`, my case is `len(opts) == 0 && req.StaleRead`.

File: distsql/request_builder.go (hunk 3 above)
Comment (Yisaer) [reply distsql/request_builder.go:240]: I think it is unnecessary, `build.txnScope` is also responsible for `(builder *RequestBuilder) verifyTxnScope()`.

File: store/tikv/client_helper.go (hunk 4 above)
Comment (Yisaer) [reply store/tikv/client_helper.go]: I think `len(opts) == 0 && req.StaleRead` is valid, it indicates a staleness transaction with global txnScope.

**Review by djshow832 — APPROVED** [review]

**Pull request merged**